File: android/app/src/main/kotlin/com/fortis/wallet/wallet/Entropy.kt
package com.fortis.wallet.wallet

import java.io.ByteArrayOutputStream
import kotlin.math.min

/**
 * Supplementary entropy for seed generation. `SecureRandom` already gives 256
 * bits and is sound on modern Android, but a wallet is worth hedging: whatever
 * is gathered here is *mixed* with the CSPRNG bytes inside wallet-ffi
 * (`wallet_core::entropy`) — it can only strengthen the seed, never weaken it,
 * so collection is best-effort and the user can skip.
 *
 * The novel bit for a phone: raw motion-sensor samples while the user shakes it.
 * The low bits of MEMS accelerometer / gyroscope readings are genuine thermal
 * and mechanical noise, and a human's chaotic shake piles more on top.
 */
class EntropyCollector {
    private val buf = ByteArrayOutputStream()
    private var estBits = 0.0

    /** Conservative lower-bound estimate of bits gathered. */
    val bits: Int get() = estBits.toInt()

    /** One sensor reading (accelerometer m/s² or gyroscope rad/s, 3 axes). */
    @Synchronized
    fun addMotion(values: FloatArray) {
        for (v in values) {
            val bits = java.lang.Float.floatToRawIntBits(v)
            buf.write(bits and 0xff)
            buf.write((bits ushr 8) and 0xff)
        }
        buf.write((System.nanoTime() and 0xff).toInt())
        estBits += 2.0 // ~2 bits/sample, conservative
    }

    /** One point of a touch drag: coordinates + event time. */
    @Synchronized
    fun addTouch(x: Float, y: Float, eventNanos: Long) {
        val xi = x.toInt(); val yi = y.toInt()
        buf.write(xi and 0xff); buf.write((xi ushr 8) and 0xff)
        buf.write(yi and 0xff); buf.write((yi ushr 8) and 0xff)
        val t = eventNanos
        buf.write((t and 0xff).toInt()); buf.write(((t ushr 8) and 0xff).toInt())
        estBits += 3.0
    }

    /** Timing jitter of a tight loop — CPU scaling / scheduling noise. Runs with
     *  no user interaction. */
    @Synchronized
    fun addJitter(iterations: Int = 4000) {
        var acc = System.nanoTime()
        var last = acc
        repeat(iterations) {
            var s = 0.0
            for (i in 0 until 200) s += Math.sqrt((i + 1).toDouble())
            val now = System.nanoTime()
            buf.write(((now - last) and 0xff).toInt())
            buf.write((s.toRawBits() and 0xff).toInt())
            last = now
            acc += now
        }
        estBits += iterations * 0.2
    }

    @Synchronized
    fun bytes(): ByteArray = buf.toByteArray()
}

/** Target extra bits before the meter reads "full" (the base seed is already
 *  256-bit; this is headroom). */
const val ENTROPY_TARGET_BITS = 128

fun entropyProgress(bits: Int): Float = min(1f, bits / ENTROPY_TARGET_BITS.toFloat())

File: web/src/entropy.js
// Supplementary entropy for seed generation.
//
// `crypto.getRandomValues` already gives 256 bits and is almost certainly sound.
// This is defence in depth: whatever is collected here is *mixed* with the
// CSPRNG bytes inside wasm (`wallet_core::entropy`) — it can only strengthen the
// seed, never weaken it, so every source is best-effort and skippable.

export class EntropyPool {
  constructor() {
    this._chunks = [];
    this._bits = 0;
  }

  /** A conservative lower-bound estimate of bits collected so far. */
  get bits() {
    return Math.floor(this._bits);
  }

  _add(bytes, bits) {
    this._chunks.push(bytes);
    this._bits += bits;
  }

  /**
   * Sample the low bits of a high-resolution timer inside a tight loop. The
   * spacing jitters with CPU frequency scaling, cache state, and OS scheduling.
   * Weak per sample (~0.5 bit) — a hedge that runs with no user interaction.
   */
  async collectJitter(ms = 300) {
    const out = [];
    const deadline = performance.now() + ms;
    let last = performance.now();
    while (performance.now() < deadline) {
      let acc = last;
      for (let i = 0; i < 700; i++) acc = Math.sqrt(acc * 1.0000001 + i);
      const now = performance.now();
      // fractional nanoseconds of the gap
      out.push(Math.floor((((now - last) % 1) + Number.EPSILON) * 1e9) & 0xff);
      out.push(Math.floor(acc) & 0xff);
      last = now;
      if (out.length % 64 === 0) await new Promise((r) => setTimeout(r, 0));
    }
    this._add(Uint8Array.from(out), out.length * 0.25);
  }

  /** One point of a human pointer/touch path: coordinates + event timestamp. */
  addPointer(x, y, t) {
    const b = new Uint8Array(10);
    const dv = new DataView(b.buffer);
    dv.setUint16(0, (x | 0) & 0xffff);
    dv.setUint16(2, (y | 0) & 0xffff);
    dv.setFloat32(4, t); // sub-ms precision where the browser allows it
    dv.setUint16(8, (performance.now() * 1000) & 0xffff);
    this._add(b, 3); // ~3 bits/sample, conservative for a chaotic drag
  }

  /** A device-motion reading (deg/s or m/s²), if the browser grants it. */
  addMotion(a, b, c) {
    const buf = new Uint8Array(12);
    new DataView(buf.buffer).setFloat32(0, a || 0);
    new DataView(buf.buffer).setFloat32(4, b || 0);
    new DataView(buf.buffer).setFloat32(8, c || 0);
    this._add(buf, 2);
  }

  /** Everything collected, as one flat Uint8Array for `generateMnemonic`. */
  bytes() {
    const n = this._chunks.reduce((s, c) => s + c.length, 0);
    const out = new Uint8Array(n);
    let o = 0;
    for (const c of this._chunks) {
      out.set(c, o);
      o += c.length;
    }
    return out;
  }
}
